### PR TITLE
planner: add simple task stack for memo

### PR DESCRIPTION
**Pull request by AilinKid**
### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tidb/issues/51664

Problem Summary:

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: pkg/planner/memo/task.go [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+
+import (
+	"strconv"
+	"sync"
+)
+
+// Task is an interface defined for all type of optimizing work: exploring, implementing, deriving-stats, join-reordering and so on.
+type Task interface {
+	// task self executing logic
+	execute()
```
Comment (qw4990) [inline pkg/planner/memo/task.go:25]: Should we add an `error` here?

Comment (AilinKid) [reply pkg/planner/memo/task.go]: make sense

File: pkg/planner/memo/task.go [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+	tasks []Task
+}
+
+func newTaskStack() *TaskStack {
+	return &TaskStack{
+		tasks: make([]Task, 0, 4),
+	}
+}
+
+func (ts *TaskStack) Len() int {
```
Comment (hawkingrei) [inline pkg/planner/memo/task.go:47]: You can add ```TaskStack.Destory``` to put it into ```pool```.

File: pkg/planner/memo/task.go [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+func newTaskStack() *TaskStack {
+	return &TaskStack{
+		tasks: make([]Task, 0, 4),
+	}
+}
+
+// Destroy indicates that when stack itself is useless like in the end of optimizing phase, we can destroy ourselves.
+func (ts *TaskStack) Destroy() {
+	// when a taskStack itself is useless, we can destroy itself actively.
+	ts.tasks = ts.tasks[:0]
```
Comment (hawkingrei) [inline pkg/planner/memo/task.go:50]: it can use clear function which is built-in

**Review by hawkingrei — APPROVED** [review]

**Pull request merged**